File: synclite-code-samples/synclite-db/python/SyncLiteDBCLient.py
import requests
import json
import os
import sys
import time
import uuid
import base64
import hashlib
import hmac

"""
* ===========================================================
  Note: 
* ===========================================================
Install following dependencies

pip install requests


This source file implements following APIs to connect to SyncLiteDB:

1. initializeDB : Initialize the given database/device of specified type (SQLITE, DUCKDB, DERBY, H2, HYPERSQL, SQLITE_APPENDER, DUCKDB_APPENDER, DERBY_APPENDER, H2_APPENDER, HYPERSQL_APPENDER, STREAMING) at the specified path. 
2. beginTransaction: Begin a transaction on specified database, returning a transaction handle
3. executeSQL: Execute specified SQL with (optional arguments for batch operations with prepared statements), on the specified database.
4. commitTransction: Commit the transaction with given transaction handle
5. rollbackTransaction: Rollback the transaction with given transaction handle
6. closeDB: Close the given database.

You can copy these APIs in your application to get started with SyncLite DB. 


The test code attempts the following operations:

1. Intialize a database of type SQLITE
2. Begin a transaction

sql: begin

3. Create a table

 sql : CREATE TABLE t1(a int, b text)

4. Insert 2 records using an INSERT prepared statement and passing a JSON array of arrays( with each inner array representing the bind paramemeters for each record) as a batch of arguments.

sql:
INSERT INTO t1 (a, b) VALUES(?, ?)

arguments: 

[
	[1, "one"]
	[2, "two]
]

5. Commit transaction:

sql: commit

6. Select records from t1

sql:
 SELECT a, b FROM t1

Response containing resultSet as a a JSON array (each record in JSON array holding a JSON object representing a table record as a map of <ColumnName, ColumnValue> pairs
   [
        {
            "a": 1,
            "b": "one"
        },
        {
            "a": 2,
            "b": "two"
        }
    ]

7. Drop table t1

sql: drop table t1

8. Close database
 
"""
class SyncLiteDBResult:
    def __init__(self, result, message, result_set=None, txn_handle=None, resultset_handle=None, has_more=None, column_metadata=None):
        self.result = result
        self.message = message
        self.result_set = result_set
        self.txn_handle = txn_handle
        self.resultset_handle = resultset_handle
        self.has_more = has_more
        self.column_metadata = column_metadata


def _to_result(json_response):
    return SyncLiteDBResult(
        result=json_response.get('result'),
        message=json_response.get('message'),
        result_set=json_response.get('resultset') if 'resultset' in json_response else None,
        txn_handle=json_response.get('txn-handle') if 'txn-handle' in json_response else None,
        resultset_handle=json_response.get('resultset-handle') if 'resultset-handle' in json_response else None,
        has_more=json_response.get('has-more') if 'has-more' in json_response else None,
        column_metadata=json_response.get('resultset-metadata') if 'resultset-metadata' in json_response else None
    )

# The base URL for the API
syncLiteDBAddress = "http://localhost:5555"

def _sha256_hex(value):
    return hashlib.sha256(value.encode("utf-8")).hexdigest()

def _build_auth_headers(payload):
    headers = {'Content-Type': 'application/json'}

    token = os.getenv("SYNCLITE_DB_AUTH_TOKEN")
    if token:
        headers['X-SyncLite-Token'] = token

    app_id = os.getenv("SYNCLITE_DB_APP_ID")
    app_secret = os.getenv("SYNCLITE_DB_APP_SECRET")
    if app_id and app_secret:
        timestamp = str(int(time.time() * 1000))
        nonce = str(uuid.uuid4())
        canonical = "POST\n/\n" + timestamp + "\n" + nonce + "\n" + _sha256_hex(payload)
        signature = base64.b64encode(
            hmac.new(app_secret.encode("utf-8"), canonical.encode("utf-8"), hashlib.sha256).digest()
        ).decode("utf-8")

        headers['X-SyncLite-App-Id'] = app_id
        headers['X-SyncLite-Timestamp'] = timestamp
        headers['X-SyncLite-Nonce'] = nonce
        headers['X-SyncLite-Signature'] = signature

    return headers

# Function to send an HTTP request
def process_request(json_request):
    try:
        print(f"Request JSON: {json.dumps(json_request, indent=4)}")

        payload = json.dumps(json_request, separators=(",", ":"), ensure_ascii=False)
        headers = _build_auth_headers(payload)
        response = requests.post(syncLiteDBAddress, data=payload.encode("utf-8"), headers=headers, timeout=10)
        
        print(f"Response Code: {response.status_code}")
        
        if response.status_code in (200, 400, 401, 413):
            json_response = response.json()
            print(f"Response JSON: {json.dumps(json_response, indent=4)}")
            
            result = json_response.get('result')
            message = json_response.get('message')
            print(f"Result: {result}")
            print(f"Message: {message}")
            
            return json_response
        raise Exception(f"Failed to get a valid response from the server : {response.status_code} : {response.text}")
    except Exception as e:
        raise Exception(f"Failed to process request: {str(e)}") from e

def initialize_db(db_path, db_type, db_name, logger_config_path=None):
    try:
        json_request = {
            "db-path": str(db_path),
            "db-type": db_type,
            "db-name": db_name,
            "sql": "initialize"
        }
        if logger_config_path:
            json_request["synclite-logger-config"] = str(logger_config_path)
        
        json_response = process_request(json_request)
        
        return _to_result(json_response)
    except Exception as e:
        raise Exception(f"Failed to initialize DB: {str(db_path)} : {str(e)}")

def close_db(db_path):
    try:
        json_request = {
            "db-path": str(db_path),
            "sql": "close"
        }

        json_response = process_request(json_request)
        
        return _to_result(json_response)
    except Exception as e:
        raise Exception(f"Failed to close DB: {str(db_path)} : {str(e)}")

def begin_transaction(db_path):
    try:
        json_request = {
            "db-path": str(db_path),
            "sql": "begin"
        }
        
        json_response = process_request(json_request)
        
        return _to_result(json_response)
    except Exception as e:
        raise Exception(f"Failed to begin transaction on DB: {str(db_path)} : {str(e)}")

def commit_transaction(db_path, txn_handle):
    try:
        json_request = {
            "db-path": str(db_path),
            "sql": "commit",
			"txn-handle": txn_handle
        }
        
        json_response = process_request(json_request)
        
        return _to_result(json_response)
    except Exception as e:
        raise Exception(f"Failed to commit transaction on DB: {str(db_path)} : {str(e)}")

def rollback_transaction(db_path, txn_handle):
    try:
        json_request = {
            "db-path": str(db_path),
			"txn-handle": txn_handle,
            "sql": "rollback"
        }
        
        json_response = process_request(json_request)
        
        return _to_result(json_response)
    except Exception as e:
        raise Exception(f"Failed to rollback transaction on DB: {str(db_path)} : {str(e)}")

def execute_sql(db_path, txn_handle, sql, arguments, data_format=None, include_metadata=None):
    try:
        json_request = {
            "db-path": str(db_path),
            "sql": sql
        }

        if txn_handle:
            json_request["txn-handle"] = txn_handle

        if arguments:
            json_request["arguments"] = arguments

        if data_format is not None:
            json_request["resultset-data-format"] = data_format

        if include_metadata is not None:
            json_request["resultset-include-metadata"] = "ON" if include_metadata else "OFF"
        
        json_response = process_request(json_request)
        
        return _to_result(json_response)
    except Exception as e:
        raise Exception(f"Failed to rollback transaction on DB: {str(db_path)} : {str(e)}")


def next_page(resultset_handle, resultset_pagination_size=None, data_format=None, include_metadata=None):
    try:
        json_request = {
            "request-type": "next",
            "resultset-handle": resultset_handle
        }
        if resultset_pagination_size and resultset_pagination_size > 0:
            json_request["resultset-pagination-size"] = resultset_pagination_size
        if data_format is not None:
            json_request["resultset-data-format"] = data_format
        if include_metadata is not None:
            json_request["resultset-include-metadata"] = "ON" if include_metadata else "OFF"

        json_response = process_request(json_request)
        return _to_result(json_response)
    except Exception as e:
        raise Exception(f"Failed to fetch next page for resultset-handle: {resultset_handle} : {str(e)}")


# Init db directory
db_dir = os.path.join(os.path.expanduser("~"), "synclite", "job1", "db")
os.makedirs(db_dir, exist_ok=True)

db_path = os.path.join(db_dir, "testPython.db")

# Initialize DB
print("=" * 56)
print("Executing initialize DB")
print("=" * 56)
result = initialize_db(db_path, "SQLITE", "testPython", None)
print(f"Result: {result.result}, Message: {result.message}")
print("=" * 56)

# Start a transaction
print("=" * 56)
print("Executing begin transaction")
print("=" * 56)
r = begin_transaction(db_path)
print(f"result : {r.result}")
print(f"message : {r.message}")
print(f"txn-handle: {r.txn_handle}")
txn_handle = r.txn_handle
if not r.result:
    sys.exit(1)
print("=" * 56)


# Create a Table
print("=" * 56)
print("Executing create table")
print("=" * 56)
r = execute_sql(db_path, txn_handle, "create table if not exists t1(a int, b text)", None)
print(f"result : {r.result}")
print(f"message : {r.message}")
if not r.result:
    sys.exit(1)
print("=" * 56)

# Insert Data in a table
print("=" * 56)
print("Executing insert into table")
print("=" * 56)
arguments = [
    [1, "one"],
    [2, "two"]
]
r = execute_sql(db_path, txn_handle, "insert into t1 (a,b) values(?, ?)", arguments)
print(f"result : {r.result}")
print(f"message : {r.message}")
if not r.result:
    sys.exit(1)
print("=" * 56)


# Commit Transaction
print("=" * 56)
print("Executing commit transaction")
print("=" * 56)
r = commit_transaction(db_path, txn_handle)
print(f"result : {r.result}")
print(f"message : {r.message}")
if not r.result:
    sys.exit(1)
print("=" * 56)


# Select from table (JSON format - default, records as {colName: colValue} dicts)
print("=" * 56)
print("Executing select from table (JSON format)")
print("=" * 56)
r = execute_sql(db_path, None, "select a, b from t1", None)
print(f"result : {r.result}")
print(f"message : {r.message}")
if r.column_metadata:
    print("\t".join(col['label'] for col in r.column_metadata))
current = r
while True:
    if current.result_set:
        for rec in current.result_set:
            print(f"a = {rec['a']}, b = {rec['b']}")
    if not current.has_more or not current.resultset_handle:
        break
    current = next_page(current.resultset_handle)
    if not current.result:
        print(f"next failed: {current.message}")
        sys.exit(1)
print("=" * 56)


# Select from table (DB format - records as value arrays, column order matches metadata)
print("=" * 56)
print("Executing select from table (DB format)")
print("=" * 56)
r = execute_sql(db_path, None, "select a, b from t1", None, data_format="DB", include_metadata=True)
print(f"result : {r.result}")
print(f"message : {r.message}")
if r.column_metadata:
    print("\t".join(col['label'] for col in r.column_metadata))
current = r
while True:
    if current.result_set:
        for row in current.result_set:
            print("\t".join(str(v) if v is not None else "null" for v in row))
    if not current.has_more or not current.resultset_handle:
        break
    current = next_page(current.resultset_handle, data_format="DB")
    if not current.result:
        print(f"next failed: {current.message}")
        sys.exit(1)
print("=" * 56)


# Drop table
print("=" * 56)
print("Executing drop table")
print("=" * 56)
r = execute_sql(db_path, None, "drop table t1", None)
print(f"result : {r.result}")
print(f"message : {r.message}")


# Close DB
print("=" * 56)
print("Executing close DB")
print("=" * 56)
r = close_db(db_path)
print(f"result : {r.result}")
print(f"message : {r.message}")
print("=" * 56)

File: synclite-code-samples/synclite-db/go/SyncLiteDBClient.go
package main

import (
	"bytes"
	"crypto/hmac"
	"crypto/rand"
	"crypto/sha256"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"io"
	"log"
	"net/http"
	"os"
	"path/filepath"
    "time"
)

/*
* ===========================================================
  Note: 
* ===========================================================

This source file implements following APIs to connect to SyncLiteDB:

1. initializeDB : Initialize the given database/device of specified type (SQLITE, DUCKDB, DERBY, H2, HYPERSQL, SQLITE_APPENDER, DUCKDB_APPENDER, DERBY_APPENDER, H2_APPENDER, HYPERSQL_APPENDER, STREAMING) at the specified path. 
2. beginTransaction: Begin a transaction on specified database, returning a transaction handle
3. executeSQL: Execute specified SQL with (optional arguments for batch operations with prepared statements), on the specified database.
4. commitTransction: Commit the transaction with given transaction handle
5. rollbackTransaction: Rollback the transaction with given transaction handle
6. closeDB: Close the given database.

You can copy these APIs in your application to get started with SyncLite DB. 


The test code attempts the following operations:

1. Intialize a database of type SQLITE
2. Begin a transaction

sql: begin

3. Create a table

 sql : CREATE TABLE t1(a int, b text)

4. Insert 2 records using an INSERT prepared statement and passing a JSON array of arrays( with each inner array representing the bind paramemeters for each record) as a batch of arguments.

sql:
INSERT INTO t1 (a, b) VALUES(?, ?)

arguments: 

[
	[1, "one"]
	[2, "two]
]

5. Commit transaction:

sql: commit

6. Select records from t1

sql:
 SELECT a, b FROM t1

Response containing resultSet as a a JSON array (each record in JSON array holding a JSON object representing a table record as a map of <ColumnName, ColumnValue> pairs
   [
        {
            "a": 1,
            "b": "one"
        },
        {
            "a": 2,
            "b": "two"
        }
    ]

7. Drop table t1

sql: drop table t1

8. Close database

*/


// SyncLiteDBResult holds the result from the SyncLite DB API
type SyncLiteDBResult struct {
	Result          bool                       `json:"result"`
	Message         string                     `json:"message"`
	ResultSet       []map[string]interface{}   `json:"resultset,omitempty"`
	TxnHandle       string                     `json:"txn-handle,omitempty"`
	ResultsetHandle string                     `json:"resultset-handle,omitempty"`
	HasMore         bool                       `json:"has-more,omitempty"`
	ColumnMetadata  []map[string]interface{}   `json:"resultset-metadata,omitempty"`
}

// SyncLiteDBAddress is the base URL for the API
var syncLiteDBAddress = "http://localhost:5555"

func buildAuthHeaders(requestBody []byte) map[string]string {
	headers := map[string]string{
		"Content-Type": "application/json",
	}

	if token := os.Getenv("SYNCLITE_DB_AUTH_TOKEN"); token != "" {
		headers["X-SyncLite-Token"] = token
	}

	appID := os.Getenv("SYNCLITE_DB_APP_ID")
	appSecret := os.Getenv("SYNCLITE_DB_APP_SECRET")
	if appID != "" && appSecret != "" {
		timestamp := fmt.Sprintf("%d", time.Now().UnixMilli())
		nonceBytes := make([]byte, 16)
		_, _ = rand.Read(nonceBytes)
		nonce := fmt.Sprintf("%x", nonceBytes)

		bodyHash := sha256.Sum256(requestBody)
		canonical := fmt.Sprintf("POST\n/\n%s\n%s\n%x", timestamp, nonce, bodyHash)

		mac := hmac.New(sha256.New, []byte(appSecret))
		mac.Write([]byte(canonical))
		signature := base64.StdEncoding.EncodeToString(mac.Sum(nil))

		headers["X-SyncLite-App-Id"] = appID
		headers["X-SyncLite-Timestamp"] = timestamp
		headers["X-SyncLite-Nonce"] = nonce
		headers["X-SyncLite-Signature"] = signature
	}

	return headers
}

// processRequest sends an HTTP request to the SyncLiteDB API
func processRequest(jsonRequest map[string]interface{}) (SyncLiteDBResult, error) {
	var result SyncLiteDBResult
	requestBody, err := json.MarshalIndent(jsonRequest, "", "  ")
	if err != nil {
		return result, fmt.Errorf("failed to marshal JSON request: %v", err)
	}

	fmt.Println("Request JSON:", string(requestBody))

	req, err := http.NewRequest("POST", syncLiteDBAddress, bytes.NewBuffer(requestBody))
	if err != nil {
		return result, fmt.Errorf("failed to create request: %v", err)
	}
	for k, v := range buildAuthHeaders(requestBody) {
		req.Header.Set(k, v)
	}

	client := &http.Client{}
	resp, err := client.Do(req)
	if err != nil {
		return result, fmt.Errorf("failed to send request: %v", err)
	}
	defer resp.Body.Close()

	fmt.Println("Response Code:", resp.StatusCode)

	if resp.StatusCode == 200 || resp.StatusCode == 400 || resp.StatusCode == 401 || resp.StatusCode == 413 {
		body, _ := io.ReadAll(resp.Body)
		fmt.Println("Response JSON:", string(body))
		err = json.Unmarshal(body, &result)
		if err != nil {
			return result, fmt.Errorf("failed to parse JSON response: %v", err)
		}
	} else {
		body, _ := io.ReadAll(resp.Body)
		return result, fmt.Errorf("failed to get a valid response from the server: %v : %s", resp.StatusCode, string(body))
	}

	return result, nil
}

// initializeDB initializes the SyncLite database
func initializeDB(dbPath, dbType, dbName, loggerConfigPath string) (SyncLiteDBResult, error) {
	jsonRequest := map[string]interface{}{
		"db-path": dbPath,
		"db-type": dbType,
		"db-name": dbName,
		"sql":     "initialize",
	}

	if loggerConfigPath != "" {
		jsonRequest["synclite-logger-config"] = loggerConfigPath
	}

	return processRequest(jsonRequest)
}

// closeDB closes the SyncLite database
func closeDB(dbPath string) (SyncLiteDBResult, error) {
	jsonRequest := map[string]interface{}{
		"db-path": dbPath,
		"sql":     "close",
	}
	return processRequest(jsonRequest)
}

// beginTransaction starts a new transaction
func beginTransaction(dbPath string) (SyncLiteDBResult, error) {
	jsonRequest := map[string]interface{}{
		"db-path": dbPath,
		"sql":     "begin",
	}
	return processRequest(jsonRequest)
}

// commitTransaction commits the current transaction
func commitTransaction(dbPath, txnHandle string) (SyncLiteDBResult, error) {
	jsonRequest := map[string]interface{}{
		"db-path":    dbPath,
		"txn-handle": txnHandle,
		"sql":        "commit",
	}

	return processRequest(jsonRequest)
}

// rollbackTransaction rolls back the current transaction
func rollbackTransaction(dbPath, txnHandle string) (SyncLiteDBResult, error) {
	jsonRequest := map[string]interface{}{
		"db-path":    dbPath,
		"txn-handle": txnHandle,
		"sql":        "rollback",
	}
	return processRequest(jsonRequest)
}

// executeSQL executes an SQL statement on the database
func executeSQL(dbPath, txnHandle, sql string, arguments [][]interface{}, dataFormat string, includeMetadata *bool) (SyncLiteDBResult, error) {
	jsonRequest := map[string]interface{}{
		"db-path": dbPath,
		"sql":     sql,
	}
	if txnHandle != "" {
		jsonRequest["txn-handle"] = txnHandle
	}
	if arguments != nil {
		jsonRequest["arguments"] = arguments
	}
	if dataFormat != "" {
		jsonRequest["resultset-data-format"] = dataFormat
	}
	if includeMetadata != nil {
		if *includeMetadata {
			jsonRequest["resultset-include-metadata"] = "ON"
		} else {
			jsonRequest["resultset-include-metadata"] = "OFF"
		}
	}

	return processRequest(jsonRequest)
}

func next(resultsetHandle string, resultsetPaginationSize int, dataFormat string, includeMetadata *bool) (SyncLiteDBResult, error) {
	jsonRequest := map[string]interface{}{
		"request-type":     "next",
		"resultset-handle": resultsetHandle,
	}
	if resultsetPaginationSize > 0 {
		jsonRequest["resultset-pagination-size"] = resultsetPaginationSize
	}
	if dataFormat != "" {
		jsonRequest["resultset-data-format"] = dataFormat
	}
	if includeMetadata != nil {
		if *includeMetadata {
			jsonRequest["resultset-include-metadata"] = "ON"
		} else {
			jsonRequest["resultset-include-metadata"] = "OFF"
		}
	}

	return processRequest(jsonRequest)
}

func main() {
	// Get the DB directory
	homeDir, _ := os.UserHomeDir()
	dbDir := filepath.Join(homeDir, "synclite", "job1", "db")
	err := os.MkdirAll(dbDir, os.ModePerm)
	if err != nil {
		log.Fatalf("Failed to create directory: %v", err)
	}

	dbPath := filepath.Join(dbDir, "testGo.db")

	// Initialize DB
	fmt.Println("========================================================")
	fmt.Println("Executing initialize DB")
	fmt.Println("========================================================")
	result, err := initializeDB(dbPath, "SQLITE", "testGo", "")
	if err != nil {
		log.Fatalf("Failed to initialize DB: %v", err)
	}
	fmt.Printf("Result: %v, Message: %v\n", result.Result, result.Message)
	fmt.Println("========================================================")

	// Begin a transaction
	fmt.Println("========================================================")
	fmt.Println("Executing begin transaction")
	fmt.Println("========================================================")
	r, err := beginTransaction(dbPath)
	if err != nil {
		log.Fatalf("Failed to begin transaction: %v", err)
	}
	fmt.Printf("result: %v, message: %v, txn-handle: %v\n", r.Result, r.Message, r.TxnHandle)
	txnHandle := r.TxnHandle
	if !r.Result {
		os.Exit(1)
	}
	fmt.Println("========================================================")

	// Create a Table
	fmt.Println("========================================================")
	fmt.Println("Executing create table")
	fmt.Println("========================================================")
	r, err = executeSQL(dbPath, txnHandle, "create table if not exists t1(a int, b text)", nil, "", nil)
	if err != nil {
		log.Fatalf("Failed to create table: %v", err)
	}
	fmt.Printf("result: %v, message: %v\n", r.Result, r.Message)
	if !r.Result {
		os.Exit(1)
	}
	fmt.Println("========================================================")

	// Insert Data in a table
	fmt.Println("========================================================")
	fmt.Println("Executing insert into table")
	fmt.Println("========================================================")
	arguments := [][]interface{}{
		{1, "one"},
		{2, "two"},
	}
	r, err = executeSQL(dbPath, txnHandle, "insert into t1 (a,b) values(?, ?)", arguments, "", nil)
	if err != nil {
		log.Fatalf("Failed to insert into table: %v", err)
	}
	fmt.Printf("result: %v, message: %v\n", r.Result, r.Message)
	if !r.Result {
		os.Exit(1)
	}
	fmt.Println("========================================================")

	// Commit Transaction
	fmt.Println("========================================================")
	fmt.Println("Executing commit transaction")
	fmt.Println("========================================================")
	r, err = commitTransaction(dbPath, txnHandle)
	if err != nil {
		log.Fatalf("Failed to commit transaction: %v", err)
	}
	fmt.Printf("result: %v, message: %v\n", r.Result, r.Message)
	if !r.Result {
		os.Exit(1)
	}
	fmt.Println("========================================================")

	// Select from table (JSON format - default, records as map[colName]colValue)
	fmt.Println("========================================================")
	fmt.Println("Executing select from table (JSON format)")
	fmt.Println("========================================================")
	r, err = executeSQL(dbPath, "", "select a, b from t1", nil, "", nil)
	if err != nil {
		log.Fatalf("Failed to select from table: %v", err)
	}
	fmt.Printf("result: %v, message: %v\n", r.Result, r.Message)
	if len(r.ColumnMetadata) > 0 {
		for i, col := range r.ColumnMetadata {
			if i > 0 {
				fmt.Print("\t")
			}
			fmt.Print(col["label"])
		}
		fmt.Println()
	}
	for _, rec := range r.ResultSet {
		fmt.Printf("a = %v, b = %v\n", rec["a"], rec["b"])
	}
	for r.HasMore && r.ResultsetHandle != "" {
		r, err = next(r.ResultsetHandle, 0, "", nil)
		if err != nil {
			log.Fatalf("Failed to fetch next result page: %v", err)
		}
		if !r.Result {
			log.Fatalf("Next page call failed: %v", r.Message)
		}
		for _, rec := range r.ResultSet {
			fmt.Printf("a = %v, b = %v\n", rec["a"], rec["b"])
		}
	}
	fmt.Println("========================================================")

	// Select from table (DB format - records as []interface{} value arrays, column order matches ColumnMetadata)
	fmt.Println("========================================================")
	fmt.Println("Executing select from table (DB format)")
	fmt.Println("========================================================")
	includeMetadata := true
	dbFmtResult, err := executeSQL(dbPath, "", "select a, b from t1", nil, "DB", &includeMetadata)
	if err != nil {
		log.Fatalf("Failed to select from table (DB format): %v", err)
	}
	fmt.Printf("result: %v, message: %v\n", dbFmtResult.Result, dbFmtResult.Message)
	if len(dbFmtResult.ColumnMetadata) > 0 {
		for i, col := range dbFmtResult.ColumnMetadata {
			if i > 0 {
				fmt.Print("\t")
			}
			fmt.Print(col["label"])
		}
		fmt.Println()
	}
	for dbFmtResult.Result {
		for _, rowRaw := range dbFmtResult.ResultSet {
			// In DB format each "row" is deserialized as map[string]interface{} by Go's JSON,
			// but the values are positional by index key. Print all values in order.
			for i, col := range dbFmtResult.ColumnMetadata {
				if i > 0 {
					fmt.Print("\t")
				}
				fmt.Print(rowRaw[col["label"].(string)])
			}
			fmt.Println()
		}
		if !dbFmtResult.HasMore || dbFmtResult.ResultsetHandle == "" {
			break
		}
		dbFmtResult, err = next(dbFmtResult.ResultsetHandle, 0, "DB", nil)
		if err != nil {
			log.Fatalf("Failed to fetch next result page (DB format): %v", err)
		}
		if !dbFmtResult.Result {
			log.Fatalf("Next page call failed: %v", dbFmtResult.Message)
		}
	}
	fmt.Println("========================================================")

	// Drop table
	fmt.Println("========================================================")
	fmt.Println("Executing drop table")
	fmt.Println("========================================================")
	r, err = executeSQL(dbPath, "", "drop table t1", nil, "", nil)
	if err != nil {
		log.Fatalf("Failed to drop table: %v", err)
	}
	fmt.Printf("result: %v, message: %v\n", r.Result, r.Message)
	fmt.Println("========================================================")

	// Close DB
	fmt.Println("========================================================")
	fmt.Println("Executing close DB")
	fmt.Println("========================================================")
	r, err = closeDB(dbPath)
	if err != nil {
		log.Fatalf("Failed to close DB: %v", err)
	}
	fmt.Printf("result: %v, message: %v\n", r.Result, r.Message)
	fmt.Println("========================================================")
}
